持ちタイムを画面に表示させる

mterada1228 opened the pull request.

mterada1228 committed:
持ちタイムを固定値を画面に表示させる

mterada1228 commented on app/services/horse_stats/best_times_by_course.rb:
```diff
@@ -0,0 +1,26 @@
+        BestTime.new(horse_race2),
+        BestTime.new(horse_race3),
+        BestTime.new(horse_race4)
+      ]
```
現時点ではこの部分はモックになっています。
最終的には、`horse_race` より競走馬が各コースにおいて記録した最も早いタイムを返すように実装します。

mterada1228 replied:
こちらコース毎のベストタイムを表示させるよう、ロジックを実装済みです。

ershn commented on app/services/horse_stats/best_time.rb:
```diff
@@ -0,0 +1,11 @@
+      @distance = horse_race.distance
+      @course_type = horse_race.course_type
+      @time = horse_race.time.strftime('%-M:%S.%1L')
+    end
```
ここは一つのマスターデータをベースにそれぞれ加工したデータを提供するパターンですが、それだと、全部 initialize の中で処理するより、個別のメソッド内で処理した方が読みやすいです。

mterada1228 replied:
[TimeByCourse クラス](https://github.com/mterada1228/keiba_expectation_app/blob/f3df7f29b129411a0fe65c6407305bf35d441b6f/app/services/horse_stats/time_by_course.rb) に変更しましたが、`initialize` ではなく個別のメソッドに別だししました。

ershn replied:
OK

ershn commented on app/services/horse_stats/best_time.rb:
```diff
@@ -0,0 +1,11 @@
+    def initialize(horse_race)
+      @distance = horse_race.distance
+      @course_type = horse_race.course_type
+      @time = horse_race.time.strftime('%-M:%S.%1L')
```
`strftime`はビュー処理となるので、ビューで呼び出した方が保守性が上がって、汎用性も上がります。

mterada1228 replied:
[TimeByCourse クラス](https://github.com/mterada1228/keiba_expectation_app/blob/f3df7f29b129411a0fe65c6407305bf35d441b6f/app/services/horse_stats/time_by_course.rb#L15-L17)ではフォーマットの変換は行わず、[View](https://github.com/mterada1228/keiba_expectation_app/blob/f3df7f29b129411a0fe65c6407305bf35d441b6f/app/views/horses/show/_stats.html.erb#L37) で変換を行うように修正しました。

ershn replied:
OK

ershn commented on app/services/horse_stats/best_time.rb:
```diff
@@ -0,0 +1,11 @@
+module HorseStats
+  class BestTime
```
このクラスで表現されてるのはベストタイムではなく、ホースの一つのタイムなので、クラス名を修正した方がいいと思います。

mterada1228 replied:
`TimeByCourse` クラスに名称を変更しました。

ershn replied:
OK

ershn commented on spec/services/horse_stats/evaluate_horse_spec.rb:
```diff
@@ -0,0 +1,28 @@
+
+  describe 'call' do
+    it '競走馬のコースごとの持ちタイムを表示する' do
+      best_times = described_class.new(horse.horse_races).call.best_times_by_course.best_times
```
`described_class.new(horse.horse_races).call`の部分を`subject`にしてもいいと思いました。

mterada1228 replied:
[テスト](https://github.com/mterada1228/keiba_expectation_app/blob/f3df7f29b129411a0fe65c6407305bf35d441b6f/spec/services/horse_stats/evaluate_horse_spec.rb#L20) は `subject` で `described_class.new(horse.horse_races).call` を行うよう修正しました。

ershn replied:
OK

mterada1228 committed:
指摘対応
  - time のフォーマット変換は View で行う
  - BestTime クラスの名称変更
ベストタイム表示ロジックを実装

ershn commented on app/services/horse_stats/best_times_by_course.rb:
```diff
@@ -0,0 +1,24 @@
+      group_by_course(horse_races).values.map do |horse_races_by_course|
+        TimeByCourse.new(horse_races_by_course.min_by(&:time))
+      end
+    end
```
このメソッドを private にしてもいいと思います。

mterada1228 replied:
private メソッドに変更しました

ershn replied:
OK

ershn commented on app/services/horse_stats/time_by_course.rb:
```diff
@@ -0,0 +1,19 @@
+module HorseStats
+  class TimeByCourse
```
`By`が入ると、連想配列だと解釈してしまうことが一般的だと思いますので、ここは`CourseTime`の方が適切だと思います。

mterada1228 replied:
`CourseTime` クラスに変更しました

ershn replied:
OK

ershn commented on spec/services/horse_stats/best_times_by_course_spec.rb:
```diff
@@ -0,0 +1,61 @@
+    end
+
+    it '最も早いタイムのみが取得できる' do
+      expect(subject.times.first.time).to eq(horse_race_record_fastest_time.time)
```
ここは以下のようにテストした方が堅牢だと思います。

```ruby
expect(subject.times).to eq([horse_race_record_fastest_time])
```

mterada1228 replied:
`expect(subject.times.map(&:time)).to eq([horse_race_record_fastest_time.time])` で検証するよう変更しました。

ershn replied:
OK

ershn commented on spec/services/horse_stats/evaluate_horse_spec.rb:
```diff
@@ -0,0 +1,42 @@
+        .to have_attributes(distance: 1800,
+                            course_type: 'dirt',
+                            time: horse_race_distanse_1800_dirt.time)
+    end
```
以下のようにもっと単純に書けるかと思います。

```ruby
expect(subject.best_times_by_course.times).to eq([TimeByCourse.new(horse_race_distanse_1000_turf), ...])
```

mterada1228 replied:
ご指摘の通り修正をしますと、オブジェクトIDが異なるためテストが失敗してしまいますので一旦修正保留しております。

ershn replied:
デフォルトで ruby のオブジェクトはアイデンティティで比較されるが、ここはバリューで比較したいので、`TimeCourse` には`==`を定義する必要があります。

そうすれば、コメントのように実装できるようになると思います。

mterada1228 committed:
指摘対応
  - BestTimesByCourse#create_stats は private メソッドに変更
  - TimeByCourse は CourseTime にクラス名を変更

mterada1228 committed:
指摘対応
